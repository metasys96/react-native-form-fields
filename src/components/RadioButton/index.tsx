/* eslint-disable react-native/no-inline-styles */
import React, {useState} from 'react';
import {View, Text, TouchableOpacity,StyleSheet,StyleProp,ViewStyle,TextStyle} from 'react-native';
import {Icon} from 'react-native-elements';
interface AppProps {
  items: Array<itemsObj>;
  size?: number;
  isVertical?: boolean;
  selectedValue?: itemsObj;
  onChange?: (value:itemsObj )=>void;
  containerStyle?:  StyleProp<ViewStyle>;
  iconContainerStyle?:  StyleProp<ViewStyle>;
  labelContainerStyle?:  StyleProp<ViewStyle>;
  labelStyle?: StyleProp<TextStyle>;
  labelFontSize?: number;
  labelColor?: string;
  selectedIcon?: selectedIconObj;
  unSelectedIcon?: unSelectedIconObj;
  selectedElement?: boolean;
  unSelectedElement?: boolean;
}

const RadioButton = (props: AppProps) => {
  const {
    items,
    size,
    isVertical,
    selectedValue,
    onChange,
    containerStyle,
    iconContainerStyle,
    labelContainerStyle,
    labelStyle,
    labelFontSize,
    labelColor,
    selectedIcon,
    unSelectedIcon,
    selectedElement,
    unSelectedElement,
  } = props;

  const [radioSelected, setRadio] = useState(
    selectedValue ? selectedValue.id : items[0].id,
  );

  const onRadioClick = (item: any) => {
    setRadio(item.id);
    if ('function' === typeof onChange) {
      onChange(item);
    }
  };

  const getOptions = () => {
    const options = items.map((item: any) => {
      return (
        <TouchableOpacity        
          key={item.id}         
          activeOpacity={1}
          onPress={() => onRadioClick(item)}
          style={[
            DefaultStyles.containerStyle,
            {height: size,marginRight: isVertical ? '0%' : '5%'},
            containerStyle,
          ]}>
          <View
            style={[
              DefaultStyles.iconContainerStyle,
              iconContainerStyle,
            ]}>
            {item.id === radioSelected ? (
              selectedElement ? (
                selectedElement
              ) : (
                <Icon
                  size={size ? size : selectedIcon.size}
                  name={selectedIcon.name}
                  type={selectedIcon.type}
                  color={selectedIcon.color}
                />
              )
            ) : unSelectedElement ? (
              unSelectedElement
            ) : (
              <Icon
                size={size ? size : unSelectedIcon.size}
                name={unSelectedIcon.name}
                type={unSelectedIcon.type}
                color={unSelectedIcon.color}
              />
            )}
          </View>
          <View style={[{justifyContent: 'center'}, labelContainerStyle]}>
            <Text
              style={[
                {
                  fontSize: size ? size * 0.8 : labelFontSize,
                  paddingLeft: '2%',
                  color: labelColor,
                },
                labelStyle,
              ]}>
              {item.value}
            </Text>
          </View>
        </TouchableOpacity>
      );
    });

    return options;
  };

  return (
    <View
      style={{
        flexDirection: isVertical ? 'column' : 'row',
        marginVertical: '1%',
      }}>
      {getOptions()}
    </View>
  );
};
const DefaultStyles = StyleSheet.create({
  iconContainerStyle:{
    justifyContent: 'center',
    shadowColor: 'rgba(0,0,0, .2)',
    shadowOffset: {height: 1, width: 1},
    shadowOpacity: 1,
    shadowRadius: 1,
  },
  containerStyle:{
    flexDirection: 'row',
    justifyContent: 'center',
  }
})


RadioButton.defaultProps = {
  items: [
    {
      id: 1,
      value: 'Option 1',
    },
    {
      id: 2,
      value: 'Option 2',
    },
  ],

  labelColor: 'black',
  isVertical: false,
  labelFontSize: 20,
  selectedIcon: {
    size: 20,
    name: 'radio-button-checked',
    type: 'MaterialIcons',
    color: 'black',
  },
  unSelectedIcon: {
    size: 20,
    name: 'radio-button-unchecked',
    type: 'MaterialIcons',
    color: 'black',
  },
};
export default RadioButton;
