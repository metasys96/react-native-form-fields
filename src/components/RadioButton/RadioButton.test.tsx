import {shallow, mount} from 'enzyme';
import RadioButton from './index';
import React from 'react';
import 'react-native';
import {cleanup, fireEvent, render} from 'react-native-testing-library';
import renderer from 'react-test-renderer';
import {View, Text, TouchableOpacity} from 'react-native';

afterEach(cleanup);

const radioOptions = [
  {id: 'option1', value: 'Red'},
  {id: 'option2', value: 'Green'},
  {id: 'option3', value: 'Blue'},
];

function getShallowComponent() {
  return shallow(
    <RadioButton
      objects={radioOptions}
      selectedValue={{id: 'option3', value: 'Blue'}}
    />,
  );
}

function renderComponent(onPressFunction: jest.Mock<any, any>) {
  return render(
    <RadioButton
      objects={radioOptions}
      selectedValue={{id: 'option3', value: 'Blue'}}
      onChange={onPressFunction}
    />,
  );
}

function getJSONRenderer() {
  return renderer.create(<RadioButton />).toJSON();
}

test('create snapshot', () => {
  const component = getJSONRenderer();
  expect(component).toMatchSnapshot();
});

it('renders component', () => {
  const component = getJSONRenderer();
  expect(component).toBeTruthy();
});

it('should contain TouchableOpacity and text as same as length of the options', () => {
  const component = getShallowComponent();
  expect(component.find('TouchableOpacity')).toHaveLength(radioOptions.length);
  expect(component.find('Text').length).toEqual(radioOptions.length);
});

it('should change the radio option on press function', async () => {
  const onPressFunction = jest.fn();
  const wrapper = shallow(
    <RadioButton
      objects={radioOptions}
      selectedValue={{id: 'option3', value: 'Blue'}}
      onChange={onPressFunction}
      size={20}
    />,
  );

  const {getAllByProps} = renderComponent(onPressFunction);
  const results = getAllByProps({objects: radioOptions});

  expect(results).toHaveLength(1);
  const radioSelectionNo = 1;
  wrapper
    .find(TouchableOpacity)
    .at(radioSelectionNo)
    .props()
    .onPress();

  expect(onPressFunction).toHaveBeenCalledTimes(1);

  expect(
    wrapper.find('TouchableOpacity').get(radioSelectionNo).props.children[0]
      .props.children.props.name,
  ).toEqual('radio-button-checked');
 
});
