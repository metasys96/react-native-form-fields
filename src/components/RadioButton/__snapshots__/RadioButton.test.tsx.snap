
exports[`create snapshot 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "marginVertical": "1%",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": "center",
        "marginRight": "5%",
        "opacity": 1,
      }
    }
    testID={1}
  >
    <View
      style={
        Array [
          Object {
            "justifyContent": "center",
            "shadowColor": "rgba(0,0,0, .2)",
            "shadowOffset": Object {
              "height": 1,
              "width": 1,
            },
            "shadowOpacity": 1,
            "shadowRadius": 1,
          },
          undefined,
        ]
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            }
          }
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
          underlayColor="white"
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "black",
                  "fontSize": 20,
                },
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="iconIcon"
          >
            
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 20,
              "paddingLeft": "2%",
            },
            undefined,
          ]
        }
      >
        Option 1
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": "center",
        "marginRight": "5%",
        "opacity": 1,
      }
    }
    testID={2}
  >
    <View
      style={
        Array [
          Object {
            "justifyContent": "center",
            "shadowColor": "rgba(0,0,0, .2)",
            "shadowOffset": Object {
              "height": 1,
              "width": 1,
            },
            "shadowOpacity": 1,
            "shadowRadius": 1,
          },
          undefined,
        ]
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            }
          }
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
          underlayColor="white"
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "black",
                  "fontSize": 20,
                },
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="iconIcon"
          >
            
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 20,
              "paddingLeft": "2%",
            },
            undefined,
          ]
        }
      >
        Option 2
      </Text>
    </View>
  </View>
</View>
`;
